appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java

```diff
@@ -629,6 +719,191 @@ protected String getRwString() {
     }
   }
 
+  /**
+   * An enum specifying the available types of helper blocks aka types of helper UI.
+   */
+  protected enum HelperType { OPTION_LIST }
+
+  /**
+   * A key that allows you to access info about a helper block.
+   */
+  protected final class HelperKey {
+    private HelperType helperType;
+
+    private String key;
+
+    /**
+     * Creates a HelperKey which can be used to access data about a helper block.
+     */
+    protected HelperKey(HelperType type, String key) {
+      this.helperType = type;
+      this.key = key;
+    }
+
+    /**
+     * Returns the type of helper block, aka the type of helper UI. Eg an option list.
+     * @return the type of helper block.
+     */
+    protected HelperType getType() {
+      return helperType;
+    }
+
+    /**
+     * Returns the key to the specific helper data. Eg in the case of an option list helper, this
+     * key could be used to look up values in the optionLists Map.
+     * @return key to the helper data.
+     */
+    protected String getKey() {
+      return key;
+    }
+  }
+
+
+  /**
+   * Represents a list of Options associated with some (enum) class. The data in this OptionList
+   * is used to create OptionList helper blocks. 
+   */
+  protected final class OptionList {
+    /**
+     * A list of option values (Strings) and option info (Options).
+     * For built-in components the Option name is used to look up the translated display text.
+     * For extensions, which do not support i18n, the Option name /is/ the display text.
+     */
+    private ArrayList<Option> options;
+
+    /**
+     * The fully qualified class name this OptionList is associated with.
+     */
+    private String className;
+
+    /**
+     * The tag name this OptionList is associated with. This goes in front of the dropdown in the
+     * blocks editor. It is usually the simplified class name.
```

Comment:
Ope actually it is always the class name! I don't think there's anything in the code that means it has to be... but I'm not 100% sure. I'll just change this to say "It is always the simplified class name."

Thanks for the catch!

Reply:
Hold on: if it's always exactly the class name, then why have it at all? There has to be a reason to justify having it. And I imagine that translation would be simplified by not having a different thing that needs to be translated. Can the notion of tag just be removed?

Reply:
Hmm, are you saying you want to remove the tag from the block? or remove the tag from the data structure? If you want to have the tag on the block I think we need store that data somewhere so that it can be passed into blockly.

Reply:
I'm probably confused, so please help to deconfuse me. I'm observing that in simple-components.json, we see the following pattern for options: 

      "className": "com.google.appinventor.components.common.Direction",
      "key": "Direction",
      "tag": "Direction",

That is, we have three fields that are very closely related: key and tag always seem to be exactly the same, and both seem to be derivable from the className as its last component. As a general rule, you don't want to have independent fields that contain similar information that must be kept in sync by the programmer when the values of some fields are automatically computable from others.  Now *maybe* I can see that there are cases where the tag name, which is presented to the App Inventor programmer, might differ from the name of the underlying Java class, Or perhaps it's just convenient not to have to perform string operations on the class name to extract the tag. But why have both key and tag if they're always the same?  Couldn't we get rid of key and always use the tag instead?

Reply:
Yeah for the current implementation we could get rid of one or the other if you want. They just represent different things, which is why I kept them separate.

The key is not held in the OptionList data type in the ComponentProcessor (as you've noticed). It is attained by the ComponentDescriptorGenerator when it goes through the optionLists_ dictionary. The key is the dictionary key. Because we can't store all of our optionLists in an optionList dictionary in simple_components.json (for backwards compatibility reasons) I decided to add the key to the option list data so that an identical dictionary could be constructed in Blockly's component_database.js.

There's no reason that the key has to be the same as the tag name. Eg we could store the fully-qualified class name as the key instead. (Maybe in the future we would want to add some functionality to do that if two simplified class names come into conflict). There's also no reason that the tag has to be the same as the simplified class name. They act completely independently, and there would be no consequences to changing one or the other.

So personally, I would prefer to keep them separate to give future work greater flexibility.

I definitely see your argument for removing one or the other though, as in the current implementation they are always the same.